lib/Platform/Intl/BCP47Parser.cpp

```diff
@@ -0,0 +1,539 @@
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alphanum{3,8};
+  return isCharType(subtagRef, 3, 8, &isASCIILetterOrDigit);
+}
+bool isTransformedExtensionKey(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alpha digit;
+  if (subtagRef.size() != 2) {
+    return false;
+  }
+  return isASCIILetter(subtagRef.front()) && isASCIIDigit(subtagRef.back());
+}
+bool isTransformedExtensionTValueItem(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (sep alphanum{3,8})+;
+  return isCharType(subtagRef, 3, 8, &isASCIILetterOrDigit);
+}
+bool isPrivateUseExtension(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (sep alphanum{1,8})+;
+  return isCharType(subtagRef, 1, 8, &isASCIILetterOrDigit);
+}
+bool isOtherExtension(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (sep alphanum{2,8})+;
+  return isCharType(subtagRef, 2, 8, &isASCIILetterOrDigit);
+}
+}
+
+class LanguageTagParser {
+ public:
+  LanguageTagParser(const std::u16string &localeId);
+  LanguageTagParser();
+  ~LanguageTagParser();
+
+  // public function declaration
+  bool parseUnicodeLocaleId();
+  ParsedLocaleIdentifier getParsedLocaleId();
+  std::u16string toString();
+
+ private:
+  // private function declaration
+  bool parseUnicodeLanguageId(bool transformedExtensionId);
+  bool addVariantSubtag(bool transformedExtensionId);
+  bool parseExtensions();
+  bool parseUnicodeExtension();
+  bool parseTransformedExtension();
+  bool parseOtherExtension(char16_t singleton);
+  bool parsePUExtension();
+  // tokenizer functions
+  std::u16string getCurrentSubtag();
+  llvh::ArrayRef<char16_t> getCurrentSubtagRef();
+  bool hasMoreSubtags();
+  bool nextSubtag();
+  
+  ParsedLocaleIdentifier parsedLocaleIdentifier;
+  std::u16string localeId_;
+  const std::u16string &originalLocaleId_;
+  int subtagStart_;
+  int subtagEnd_;
+};
+
+LanguageTagParser::LanguageTagParser(const std::u16string &localeId) : localeId_{localeId}, originalLocaleId_(localeId), subtagStart_(0), subtagEnd_(-1) {
+  toASCIILowerCase(localeId_);
+}
+LanguageTagParser::~LanguageTagParser() = default;
+
+ParsedLocaleIdentifier LanguageTagParser::getParsedLocaleId(){
+  return parsedLocaleIdentifier;
+}
+
+bool LanguageTagParser::parseUnicodeLocaleId() {
+  if (!parseUnicodeLanguageId(false))  {
+    return false;
+  }
+  if (!hasMoreSubtags()) {
+    return true;
+  }
+  if (!parseExtensions()) {
+    return false;
+  }
+  
+  return !hasMoreSubtags();
+}
+
+bool LanguageTagParser::parseUnicodeLanguageId(bool transformedExtensionId) {
+  if (!transformedExtensionId) {
+    if (!nextSubtag() || !isUnicodeLanguageSubtag(getCurrentSubtagRef())) {
```

Comment:
For every other function, we advance to the subtag we want it to process before invoking it. So it feels a little weird that this function has to initialise the subtag on its own. Perhaps it would make sense to hoist the call to `nextSubtag` out into `parseUnicodeLocaleId`?